### PR TITLE
Lowe et al 2019 - Fig 3

### PR DESCRIPTION
**Pull request by claresinger**
Building off of PR #42

### CODE REVIEW

**Comment by claresinger** [comment]
@slayoo As you can see ([here](https://github.com/claresinger/PySDM-examples/blob/lowe_fig3/PySDM_examples/Lowe_et_al_2019/aerosol_fig3.py)) the plots look totally wrong so far. But at least the skeleton for Fig 3 is here now.

One thing to note: I am getting that error “fa * fb < 0” for large values of f_org, approx > 0.8. I’m not sure why that would be. Maybe the fixes you’re working on for the Nascent aerosol case will fix this too.

**Comment by slayoo** [comment]
Thanks Clare!!!

> One thing to note: I am getting that error “fa * fb < 0” for large values of f_org, approx > 0.8. I’m not sure why that would be. Maybe the fixes you’re working on for the Nascent aerosol case will fix this too.

Yes, this is very likely the same problem. A potential offender here is the sgm_w in critical radius evaluation for picking the root search interval in wet radius initialisation:
https://github.com/slayoo/PySDM/blob/compressed_film/PySDM/initialisation/r_wet_init.py#L54

**Comment by slayoo** [comment]
@claresinger, happy to help here, please could you summarize the status and where to start from?

**Comment by claresinger** [comment]
@slayoo Thanks for your help.

So far the changes I've made are minor. Basically adding a few extra keywords to the settings and aerosol with default values so that I can run the parcel model with different updraft velocity `w`, different accumulation mode number concentration `Acc_N2`, and different organic mass fractions in the aerosol `Acc_Forg`.

Then I've just set up the [notebook](https://github.com/claresinger/PySDM-examples/blob/lowe_fig3/PySDM_examples/Lowe_et_al_2019/fig_3.ipynb) to run a bunch of parcel models across the parameter ranges necessary to try and reproduce Fig 3 from Lowe et al. (2019). No technical difficulties to speak of, just some debugging to do to try and figure out why the resulting plots look so different from the paper. What is plotted is the percentage difference in cloud droplet number concentration between the runs with the compressed film model vs. the bulk Köhler model.

**Comment by claresinger** [comment]
Ah, thanks for catching that import statement @slayoo. I had fixed it locally but hadn't pushed yet!

**Comment by slayoo** [comment]
here's a PR to your branch with some more updates: https://github.com/claresinger/PySDM-examples/pull/1

**Comment by claresinger** [comment]
Thanks @slayoo for fixing the joblib dependencies!

**Comment by claresinger** [comment]
TODO: https://github.com/claresinger/PySDM-examples/issues/4

File: PySDM_examples/Lowe_et_al_2019/settings.py [hunk 1]
```diff
@@ -20,17 +27,27 @@ def __init__(self, dt: float, n_sd_per_mode: int,
...
         const = self.formulae.constants
         self.aerosol = aerosol
         self.spectral_sampling = spectral_sampling
-        self.t_max = (400 + 196) * si.s
-        self.output_interval = 10 * si.s
-        self.dt = dt
-
-        self.w = .32 * si.m / si.s
+        
+        max_altitude = 190 * si.m
```
Comment (slayoo) [inline PySDM_examples/Lowe_et_al_2019/settings.py:45]: @claresinger shouldn't it be 200 m ?

Comment (claresinger) [reply PySDM_examples/Lowe_et_al_2019/settings.py:45]: Well they say 200 m, but the (400+196)*0.32 = 190.72 m, like you pointed out, so I changed it just in case these 10 m mattered.

Comment (slayoo) [reply PySDM_examples/Lowe_et_al_2019/settings.py:45]: but it is `max_altitude = 190 * si.m`, not 200 here

Comment (claresinger) [reply PySDM_examples/Lowe_et_al_2019/settings.py:45]: right, I changed it recently to check the impact on the size distribution in fig 2.

**Comment by claresinger** [comment]
@slayoo getting an `int` error with `dz=0.1` on fig 3 (but not `dz=1`)
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
/var/folders/n7/2_rw_xx11dd93cgp54t6tbn80000gn/T/ipykernel_91701/2014265642.py in <module>
     10 print(updraft_list)
     11 with parallel_backend('loky', n_jobs=1):
---> 12     output = dict(Parallel(verbose=10)(
     13         delayed(compute)(subplot+f"_w{w:.2f}_f{Forg:.2f}_"+model, Settings(
     14             dz = 0.1 * si.m,

~/anaconda3/envs/pysdm/lib/python3.9/site-packages/joblib/parallel.py in __call__(self, iterable)
   1044                 self._iterating = self._original_iterator is not None
   1045 
-> 1046             while self.dispatch_one_batch(iterator):
   1047                 pass
   1048 

~/anaconda3/envs/pysdm/lib/python3.9/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)
    859                 return False
    860             else:
--> 861                 self._dispatch(tasks)
    862                 return True
    863 

~/anaconda3/envs/pysdm/lib/python3.9/site-packages/joblib/parallel.py in _dispatch(self, batch)
    777         with self._lock:
    778             job_idx = len(self._jobs)
--> 779             job = self._backend.apply_async(batch, callback=cb)
    780             # A job can complete so quickly than its callback is
    781             # called before we get here, causing self._jobs to

~/anaconda3/envs/pysdm/lib/python3.9/site-packages/joblib/_parallel_backends.py in apply_async(self, func, callback)
    206     def apply_async(self, func, callback=None):
    207         """Schedule a func to be run"""
--> 208         result = ImmediateResult(func)
    209         if callback:
    210             callback(result)

~/anaconda3/envs/pysdm/lib/python3.9/site-packages/joblib/_parallel_backends.py in __init__(self, batch)
    570         # Don't delay the application, to avoid keeping the input
    571         # arguments in memory
--> 572         self.results = batch()
    573 
    574     def get(self):

~/anaconda3/envs/pysdm/lib/python3.9/site-packages/joblib/parallel.py in __call__(self)
    260         # change the default number of processes to -1
    261         with parallel_backend(self._backend, n_jobs=self._n_jobs):
--> 262             return [func(*args, **kwargs)
    263                     for func, args, kwargs in self.items]
    264 

~/anaconda3/envs/pysdm/lib/python3.9/site-packages/joblib/parallel.py in <listcomp>(.0)
    260         # change the default number of processes to -1
    261         with parallel_backend(self._backend, n_jobs=self._n_jobs):
--> 262             return [func(*args, **kwargs)
    263                     for func, args, kwargs in self.items]
    264 

/var/folders/n7/2_rw_xx11dd93cgp54t6tbn80000gn/T/ipykernel_91701/2014265642.py in compute(key, settings)
      1 def compute(key, settings):
      2     simulation = Simulation(settings)
----> 3     output = simulation.run()
      4     output['updraft'] = settings.w
      5     output['org_fraction'] = settings.aerosol.aerosol_modes_per_cc[0]['f_org']

~/Research/PySDM-examples/PySDM_examples/Lowe_et_al_2019/simulation.py in run(self)
     93     def run(self):
     94         output = {k: [] for k in self.particulator.products}
---> 95         for step in self.settings.output_steps:
     96             self.particulator.run(step - self.particulator.n_steps)
     97             self._save_scalars(output)

~/Research/PySDM-examples/PySDM_examples/Lowe_et_al_2019/settings.py in output_steps(self)
     88     def output_steps(self) -> np.ndarray:
     89         return np.arange(0, self.nt + 1, self.steps_per_output_interval)

~/Research/PySDM-examples/PySDM_examples/Lowe_et_al_2019/settings.py in nt(self)
     79         nt = self.t_max / self.dt
     80         assert nt == int(nt)
---> 81         return int(nt)
     82 
     83     @property

AssertionError: 
```

**Comment by slayoo** [comment]
> @slayoo getting an `int` error with `dz=0.1` on fig 3 (but not `dz=1`)

I'm pulling the branch and will try it locally to figure it out...

**Comment by slayoo** [comment]
> > @slayoo getting an `int` error with `dz=0.1` on fig 3 (but not `dz=1`)
> 
> I'm pulling the branch and will try it locally to figure it out...

@claresinger please try the following:
```python
    @property
    def nt(self) -> int:
        nt = self.t_max / self.dt
        nt_int = round(nt)
        np.testing.assert_almost_equal(nt, nt_int)
        return nt_int
```

**Comment by claresinger** [comment]
> > > @slayoo getting an `int` error with `dz=0.1` on fig 3 (but not `dz=1`)
> > 
> > 
> > I'm pulling the branch and will try it locally to figure it out...
> 
> @claresinger please try the following:
> 
> ```python
>     @property
>     def nt(self) -> int:
>         nt = self.t_max / self.dt
>         nt_int = round(nt)
>         np.testing.assert_almost_equal(nt, nt_int)
>         return nt_int
> ```

This works! Thanks @slayoo

**Comment by claresinger** [comment]
Hi @jb-mackay! Clearly I don't know how this new formatter works. I tried `black PySDM-examples/Lowe_et_al_2019/*.py` locally and then pushed the changes, but it still fails the formatting check. What am I doing wrong?

**Comment by jb-mackay** [comment]
@claresinger if you haven't already, check out [this comment](https://github.com/atmos-cloud-sim-uj/PySDM/pull/763#issuecomment-1059604900) which explains how to set up the pre-commit hooks locally so that they automatically run on any files you have changed (or apply to all files).

It looks like 3 files are still being formatted:
reformatted PySDM_examples/Lowe_et_al_2019/aerosol.py (by black)
reformatted PySDM_examples/Lowe_et_al_2019/settings.py (by black)
Fixing PySDM-examples/PySDM_examples/Lowe_et_al_2019/plot_helper.py (by isort)

If you run black/isort on these locally is it still making changes? And if so, what do they look like? We had some problems early on where black & isort fought each other, but that should not be the case given the pyproject.toml config.

@slayoo this may be a reason to set up the token-automated commits of the formatter sometime soon.

**Comment by claresinger** [comment]
Thanks @jb-mackay! That worked great :) 

> @claresinger if you haven't already, check out [this comment](https://github.com/atmos-cloud-sim-uj/PySDM/pull/763#issuecomment-1059604900) which explains how to set up the pre-commit hooks locally so that they automatically run on any files you have changed (or apply to all files).

**Comment by slayoo** [comment]
BTW, just added the info on pre-commit hooks setup to the README:
https://github.com/atmos-cloud-sim-uj/PySDM/commit/8e45d9fea7d6d201da31c0a6b2f966cfd2fcee3b

**Comment by slayoo** [comment]
@claresinger it's green now, shall we merge?

**Comment by claresinger** [comment]
> @claresinger it's green now, shall we merge?

Talking with @rxward about this next Tuesday. We are still seemingly far from reproducing Fig 3 from the Lowe paper; our 2d contour plots looks quite different. But I feel stuck on what might be causing the differences. Will follow up next week.

**Comment by slayoo** [comment]
Thank you, Clare

**Comment by claresinger** [comment]
@rxward @slayoo - Assuming all the checks pass, I think we should merge this in. If we discover the discrepancy between what we are doing and the Lowe et al 2019 paper then we can make those changes later on, but until then it's easier if we have it merged into the main codebase so any downstream changes will be automatically incorporated.

**Comment by claresinger** [comment]
@slayoo I see this is failing because https://github.com/atmos-cloud-sim-uj/PySDM/pull/424 was closed, but is still referenced as a TODO in the Shipway and Hill example ([here](https://github.com/claresinger/PySDM-examples/blob/lowe_fig3/PySDM_examples/Shipway_and_Hill_2012/settings.py)). Should I remove those TODOs or should I link them to the new PR https://github.com/atmos-cloud-sim-uj/PySDM/pull/817?

**Comment by slayoo** [comment]
Thank you, @claresinger  and sorry for breaking it.
Yes, referencing #817 would be a best option now.

**Comment by claresinger** [comment]
The build is now failing only on windows. I think the fig 3 notebook is timing out in the test. It does take awhile to run as I have it set up (4*7*7*2=392 tasks). Should I reduce this number? Or can I increase the time limit for the cell execution on the tests? Or can I skip this test for this notebook? What do you recommend @slayoo?

**Comment by slayoo** [comment]
In general, the notebook execution on GitHub Actions is meant just to check the notebook syntax and compatibility of the code with ever-changing APIs/nuances of external packages (incl. PySDM).

Since asserts in the notebooks are not a best idea, making any long computations does not really provide any value for these CI runs of the notebooks. Thus, the practice which we often have here is to reduce number of cases or resolution within the notebook conditionally with constructs such as `n_points = 6 if 'CI' not in os.environ else 2` (example from `PySDM_examples/Jaruga_and_Pawlowska_2018/fig_2.ipynb`). This also helps reduce the CI workload which is important for three reasons:
- we intend to have more and more notebooks in PySDM-examples
- we want to get CI red or green flags ASAP in PRs
- M$ GitHub will start billing us at some point

Anything that is worth checking with asserts on actual computed values would best be added through PySDM smoke tests (like, e.g., `tests/smoke_tests/pyrcel/test_parcel_example.py`)

Thanks, Clare!

**Comment by claresinger** [comment]
Thanks for this explanation Sylwester! I have tried reducing the number of points in the `w` x `Forg` grid down to 1x1 from 3x3 on CI to hopefully alleviate this timeout error. I can also try to increase the time step only on CI if this isn't successful.

**Comment by claresinger** [comment]
@slayoo ready to merge!

**Comment by slayoo** [comment]
Thank you, Clare!
I'll make a new release soon
